bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/internal/scheduler2/SchedulerTest.java

```diff
@@ -0,0 +1,175 @@
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ *
+ * @author Peter Kriens - initial contribution and API
+ * @author Simon Kaufmann - adapted to Java 8
+ *
+ */
+public class SchedulerTest {
+    InternalSchedulerImpl si = new InternalSchedulerImpl();
+
+    @Before
+    public void setUp() {
+        si.activate();
+    }
+
+    @After
+    public void tearDown() {
+        si.deactivate();
+    }
+
+    @Test
+    public void testCronReboot() throws Exception {
+        long now = System.currentTimeMillis();
+        Semaphore s = new Semaphore(0);
+        si.schedule((foo) -> {
+            s.release();
+        }, null, "@reboot");
+        s.acquire(1);
+
+        long diff = (System.currentTimeMillis() - now);
+        assertTrue(diff < 100);
+    }
+
+    @Test
+    public void testCron2() throws Exception {
+        long now = System.currentTimeMillis();
+        AtomicReference<Object> ref = new AtomicReference<>();
+
+        Semaphore s = new Semaphore(0);
+        si.schedule((foo) -> {
+            s.release();
+            ref.set(foo.get("foo"));
+        }, Collections.singletonMap("foo", "bar"), "#\n" //
+                + "\n" //
+                + " foo = bar \n" //
+                + "# bla bla foo=foo\n" //
+                + "0/2 * * * * *");
+        s.acquire(2);
+
+        long diff = (System.currentTimeMillis() - now + 500) / 1000;
+        assertTrue(diff >= 3 && diff <= 4);
+        assertEquals("bar", ref.get());
+    }
+
+    @Test
+    public void testCancellableWithTimeout() throws InterruptedException, InvocationTargetException {
+        CompletableFuture<Integer> d = new CompletableFuture<>();
+        CompletableFuture<Integer> before = si.before(d, Duration.ofMillis(100));
+        before.cancel(true);
+        // assertEquals(CancelException.SINGLETON, before.getFailure());
+        assertTrue(before.isCompletedExceptionally());
+    }
+
+    @Test
+    public void testResolveWithTimeout() throws Exception {
+        CompletableFuture<Integer> d = new CompletableFuture<>();
+        CompletableFuture<Integer> before = si.before(d, Duration.ofMillis(100));
+        d.complete(3);
+        assertTrue(before.isDone());
+        assertEquals(Integer.valueOf(3), before.get());
+    }
+
+    @Test
+    public void testFailureWithTimeout() throws InterruptedException, InvocationTargetException {
+        CompletableFuture<Integer> d = new CompletableFuture<>();
+        CompletableFuture<Integer> before = si.before(d, Duration.ofMillis(100));
+        Exception e = new Exception();
+        d.completeExceptionally(e);
+        assertTrue(before.isDone());
+        assertTrue(before.isCompletedExceptionally());
+    }
+
+    @Test
+    public void testTimeout() throws InterruptedException, InvocationTargetException {
+        CompletableFuture<Integer> d = new CompletableFuture<>();
+        CompletableFuture<Integer> before = si.before(d, Duration.ofMillis(100));
+        Thread.sleep(200);
```

Comment:
To me it does not make complete sense, I would expect that the scheduler runs a separate thread so the only way to test a timeout is by passing in something that will take to long complete, or am I missing some essential detail here?